### PR TITLE
oci: implemented CDI device mapping

### LINKED ISSUE #1394: Add CDI --device support to --oci mode
**Is your feature request related to a problem? Please describe.**

SingularityCE doesn't currently support the new CDI standard for making hardware devices available in containers.

The `--oci` experimental runtime mode currently exposes only a naive binding approach to add devices and libraries to a container, via the `--nv` and `--oci` flags.

The `--nv` and `--oci` approach cannot support a range of functionality that is valuable, such as masking specific GPUs, introducing only subsets of device functionality into a container etc. In addition it is vendor specific, but we'd like to support e.g. Intel GPUs #1094 

**Describe the solution you'd like**

The experimental `--oci` runtime mode should support a `--device` flag, allowing the user to request a device is made available in the container using a CDI configuration.

**Additional context**

Because `--oci` mode is based on OCI runtime spec generation, and invocation of `runc` / `crun`, it should be simple to implement support via spec modification carried out in the CDI module - https://github.com/container-orchestrated-devices/container-device-interface/tree/main/pkg/cdi

### PR DESCRIPTION
**Pull request by preminger**
## Description of the Pull Request (PR):

Added a `--device` flag to the experimental `--oci` runtime mode, allowing the user to make a device available in the contained using a CDI configuration.

### This fixes or addresses the following GitHub issues:

 - Fixes #1394

### CODE REVIEW

File: cmd/internal/cli/action_flags.go
Comment (dtrudg) [inline cmd/internal/cli/action_flags.go:822]:
```suggestion
	Usage:        "a CDI device mapping. A device mapping consists of a kind specification followed by the equal sign (=) and a device label (e.g. vendor.com/device=mydevice). Kind specifications consist of a prefix and a name, separated by a slash (e.g. vendor.com/device, in the previous example). Multiple device mappings can be given by a comma separated list.",
```

File: internal/pkg/runtime/launcher/oci/launcher_linux.go
Comment (dtrudg) [inline internal/pkg/runtime/launcher/oci/launcher_linux.go:224]: Although this is a small and simple loop, please could you pull it out into a function so it ends up as `getCDIMounts` returning a slice to append. Ideally in a `cdi_linux.go` file. It's just a preference for separating stuff a bit more as this launcher code is going to see a lot of work, and expansion for 4.0.

File: internal/pkg/runtime/launcher/oci/spec_linux.go
Comment (dtrudg) [inline internal/pkg/runtime/launcher/oci/spec_linux.go:123]: Please could you move this into a new `cdi_linux.go` file here. It's just a preference for separating stuff a bit more as this launcher code is going to see a lot of work, and expansion for 4.0.

File: internal/pkg/runtime/launcher/util.go
Comment (dtrudg) [inline internal/pkg/runtime/launcher/util.go:53]: Let's bring this across to `cdi_linux.go` too. The other things here should end up getting factored out into a higher level utils pkg by 4.0.

File: internal/pkg/runtime/launcher/options.go
Comment (dtrudg) [inline internal/pkg/runtime/launcher/options.go:146]: As at the CLI layer, lets call this `Device` in case (a) we end up implementing non-CDI `--device` binding, and (b) it moves out of `internal` for 4.0... which is an aim... sensible semver controlled public API.

File: internal/pkg/runtime/launcher/oci/launcher_linux.go
Comment (dtrudg) [reply internal/pkg/runtime/launcher/oci/launcher_linux.go:224]: Actually - I see why the above isn't easy, sorry.... there's only a path to mutate the spec.

Let's go for an `addCDIDevices` function, then... which will take the spec and modify it.

Comment (dtrudg) [reply internal/pkg/runtime/launcher/oci/launcher_linux.go:224]: Yeah - per my follow-up, ignore that. I had thought it might be easier to get mounts, env vars etc out individually.. but I should have checked in the CDI code first.

File: internal/pkg/runtime/launcher/options.go
Comment (elezar) [inline internal/pkg/runtime/launcher/options.go:146]: nit: Should this be `Devices` since it is a slice?

File: internal/pkg/runtime/launcher/oci/cdi_linux.go
Comment (elezar) [inline internal/pkg/runtime/launcher/oci/cdi_linux.go:19]: nit: `addCDIDevices` since this adds multiple devices?

File: internal/pkg/runtime/launcher/oci/cdi_linux.go
Comment (elezar) [inline internal/pkg/runtime/launcher/oci/cdi_linux.go:28]: Note that the `registry.InjectDevices` function accepts a variadic list of fully-qualified device names. This includes a check for a valid name. As such I don't know whether introducing an `addSingleCDIDevice` function adds specific value.

File: internal/pkg/runtime/launcher/options.go
Comment (elezar) [inline internal/pkg/runtime/launcher/options.go:500]: Does it make sense to call `IsQualifiedName` for each of the elements of `op` here to validate the options. Then it should not be required to perform this check explicitly later and the check for further validity can be delegated to the registry.

File: cmd/internal/cli/action_flags.go
Comment (elezar) [inline cmd/internal/cli/action_flags.go:822]: I think the terms we use are `fully-qualified CDI device name` instead of `mapping`. The parts are:

`<VENDOR>/<CLASS>=<NAME>` and together these are a "fully-qualified CDI device name".

File: cmd/internal/cli/action_flags.go
Comment (elezar) [inline cmd/internal/cli/action_flags.go:816]: This comment seems incorrect.

File: cmd/internal/cli/action_flags.go
Comment (elezar) [inline cmd/internal/cli/action_flags.go:40]: `devices`?

**Review by elezar — COMMENTED** [review]
Some comments from the perspective of CDI. Here I would obviously defer to singulartiy maintainers.

One thing that I was wonder was whether we want to make the use of CDI devices mutually exclusive with other mechanisms for injecting NVIDIA devices? I would argue that this would make sense sicne using them both may lead to hard-to-diagnose problems due to the interactions with the `nvidia-container-runtime-hook`. This could be implemented as a follow-up though.

File: internal/pkg/runtime/launcher/options.go
Comment (dtrudg) [reply internal/pkg/runtime/launcher/options.go:500]: Not really... this could be open to non-CDI `--device` usage in future, and as there's an intention to move the `Opt*` and stuff out of `internal`, for public consumption... I'd rather keep CDI specific stuff separate at present.

File: cmd/internal/cli/action_flags.go
Comment (dtrudg) [reply cmd/internal/cli/action_flags.go:40]: Definitely just `device` here at present please. We are matching the existing pattern for other flag vars where the var is named the same as the flag, which is non-plural `--device`.

If we change it to plural (for which I see the argument, and tend to agree), we'll do that consistently across our other flag vars too, in a separate re-factoring.

**Comment by dtrudg** [comment]
> One thing that I was wonder was whether we want to make the use of CDI devices mutually exclusive with other mechanisms for injecting NVIDIA devices? I would argue that this would make sense sicne using them both may lead to hard-to-diagnose problems due to the interactions with the `nvidia-container-runtime-hook`.

I don't think we want to do this, personally... other than perhaps issuing a warning. I'd only support making them mutually exclusive if there's a way to scope it easily to *only* NVIDIA CDI devices.

I can definitely foresee situations where someone might be on a system with no NVIDIA CDI configuration in place yet, but some other CDI configuration for another device. We see the most unlikely combinations of config on hosts, quite regularly :-)

**Comment by elezar** [comment]
> > One thing that I was wonder was whether we want to make the use of CDI devices mutually exclusive with other mechanisms for injecting NVIDIA devices? I would argue that this would make sense sicne using them both may lead to hard-to-diagnose problems due to the interactions with the `nvidia-container-runtime-hook`.
> 
> I don't think we want to do this, personally... other than perhaps issuing a warning. I'd only support making them mutually exclusive if there's a way to scope it easily to _only_ NVIDIA CDI devices.
> 
> I can definitely foresee situations where someone might be on a system with no NVIDIA CDI configuration in place yet, but some other CDI configuration for another device. We see the most unlikely combinations of config on hosts, quite regularly :-)

You're right. I forgot about the more general use cases here. Sorry about that.

**Review by dtrudg — APPROVED** [review]
This looks good to me to merge as the initial CDI implementation.

If there are any additional things you want to do, please ask @wobito or another to re-review while I'm travelling.

We wil handle a bit more e2e-testing later on. There are the GPU tests, in which we can do some stuff to actually test whether an Nvidia GPU is visible in the container... on systems with one of those :-)

**Pull request merged**
